### PR TITLE
Alerting: Support UTF-8

### PR DESCRIPTION
**Pull request by grobinson-grafana**
**What is this feature?**

This pull request updates our fork of Alertmanager to commit [65bdab0](https://github.com/grafana/prometheus-alertmanager/commit/65bdab09d3ed64c7bdd9474f19d4d48f857240fb), which is based on commit [5658f8c](https://github.com/prometheus/alertmanager/commit/5658f8cfb59c9e23bcfed3bc66cd47a7b6d92a85) in Prometheus Alertmanager.

It applies the changes from [grafana/alerting#155](https://github.com/grafana/alerting/pull/155) which removes the overrides for validation of alerts, labels and silences that we had put in place to allow alerts and silences to work for non-Prometheus datasources. However, as this is now supported in Alertmanager with the UTF-8 work, we can use the new upstream functions and remove these overrides.

The compat package is a package in Alertmanager that takes care of backwards compatibility when parsing matchers, validating alerts, labels and silences. The package has three modes:

1. Classic mode, which is Alertmanager with Prometheus datasource restrictions.
2. UTF-8 strict mode, which is UTF-8 with no backwards compatibility.
3. Fallback mode, which is a combination of the two previous modes, where Alertmanager first checks if its UTF-8 compatible, and if not, falls back to classic mode.

These modes are controlled via `compat.InitFromFlags`. Grafana initializes the compat package without any feature flags, which is the equivalent of fallback mode. Classic and UTF-8 strict mode are used in Mimir.

While Grafana Managed Alerts have no need for fallback mode, Grafana can still be used as an interface to manage the configurations of Mimir Alertmanagers and view configurations of Prometheus Alertmanager, and those installations might not have migrated or being running on older versions. Such installations behave as if in classic mode, and Grafana must be able to parse their configurations to interact with them for some period of time. As such, Grafana uses fallback mode until we are ready to drop support for outdated installations of Mimir and the Prometheus Alertmanager.

Fallback mode does not affect Grafana Managed Alerts for the following reasons:

1. Grafana uses `object_matchers` instead of `matchers` in its Alertmanager configuration file.
2. Even if users do use `matchers`, UTF-8 in `matchers` is still supported as normal. Any matchers incompatible with the UTF-8 parser will fallback to classic mode and a warning will be logged by Grafana.
3. Silences are created using structured JSON, such that no matchers need to be parsed.
4. In fallback mode, any UTF-8 string is a valid label, so we don't need to worry about breaking rules that use non Prometheus datasources, or breaking the creation and management of silences.
5. Inhibition rules are not supported. And even if users have somehow created inhibition rules, any inhibition rules that contain UTF-8 are still supported as normal. Any inhibition rules incompatible with the UTF-8 parser will fallback to classic mode and a warning will be logged by Grafana.

You may have seen that the compat package is initialized in `grafana/alerting`. It is re-initialized again in the `ngalert` package so we can use the standard logger. The package - as initialized in `grafana/alerting` - does not have a logger available.

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by grobinson-grafana** [comment]
The following screenshots show an example of UTF-8 in Grafana Managed Alerts. This is not new behavior - instead - it's to demonstrate that this PR has no breaking changes.

I created an alert rule with two custom labels `foo🙂=bar` and `bar🙂=baz`:

Screenshot 2024-02-02 at 1 05 02 PM

Here you can see the alert firing for this rule:

Screenshot 2024-02-02 at 1 05 13 PM

Here is the same alert in the Groups tab:

Screenshot 2024-02-02 at 1 05 39 PM

I created a notification policy that matches these labels and delivers notification to a webhook:

Screenshot 2024-02-02 at 1 05 29 PM

Here is the output of the webhook:

```
{"receiver":"Test","status":"firing","alerts":[{"status":"firing","labels":{"alertname":"Test","bar":"baz🙂","foo🙂":"bar","grafana_folder":"Test Folder"},"annotations":{},"startsAt":"2024-02-02T13:04:00Z","endsAt":"0001-01-01T00:00:00Z","generatorURL":"http://localhost:3000/alerting/grafana/cdbkpp5xuafpce/view?orgId=1","fingerprint":"c1e06109f91d8bb3","silenceURL":"http://localhost:3000/alerting/silence/new?alertmanager=grafana\u0026matcher=alertname%3DTest\u0026matcher=bar%3Dbaz%F0%9F%99%82\u0026matcher=foo%F0%9F%99%82%3Dbar\u0026matcher=grafana_folder%3DTest+Folder\u0026orgId=1","dashboardURL":"","panelURL":"","values":{"A":1,"B":1},"valueString":"[ var='A' labels={} value=1 ], [ var='B' labels={} value=1 ]"}],"groupLabels":{"alertname":"Test","grafana_folder":"Test Folder"},"commonLabels":{"alertname":"Test","bar":"baz🙂","foo🙂":"bar","grafana_folder":"Test Folder"},"commonAnnotations":{},"externalURL":"http://localhost:3000/","version":"1","groupKey":"{}/{bar=\"baz🙂\",foo🙂=\"bar\"}:{alertname=\"Test\", grafana_folder=\"Test Folder\"}","truncatedAlerts":0,"orgId":1,"title":"[FIRING:1] Test Test Folder (baz🙂 bar)","state":"alerting","message":"**Firing**\n\nValue: A=1, B=1\nLabels:\n - alertname = Test\n - bar = baz🙂\n - foo🙂 = bar\n - grafana_folder = Test Folder\nAnnotations:\nSource: http://localhost:3000/alerting/grafana/cdbkpp5xuafpce/view?orgId=1\nSilence: http://localhost:3000/alerting/silence/new?alertmanager=grafana\u0026matcher=alertname%3DTest\u0026matcher=bar%3Dbaz%F0%9F%99%82\u0026matcher=foo%F0%9F%99%82%3Dbar\u0026matcher=grafana_folder%3DTest+Folder\u0026orgId=1\n"}
```

I then silenced the alert:

Screenshot 2024-02-02 at 1 14 37 PM

**Comment by grobinson-grafana** [comment]
The following screenshots show the Grafana UI managing a Grafana Cloud Alertmanager. This Alertmanager has the following matchers:

```yml
"matchers": [
    "foo=",
    "bar=baz",
    "baz=\"\\xf0\\x9f\\x99\\x82\""
],
```

However, `foo=` is incompatible with UTF-8 strict mode because it is missing a value, and `baz="\xf0\x9f\x99\x82"` causes disagreement because in UTF-8 strict mode it is an 🙂, but parses as a literal `\xf0\x9f\x99\x82` in classic mode because the old parser does not understand such escape sequences.

The screenshot shows that, in fallback mode, Grafana can still read and show such configurations:

Screenshot 2024-02-02 at 1 17 50 PM

and that the following warnings are logged:

```
WARN [02-02|13:18:27] Alertmanager is moving to a new parser for labels and matchers, and this input is incompatible. Alertmanager has instead parsed the input using the old matchers parser as a fallback. To make this input compatible with the new parser please make sure all regular expressions and values are double-quoted. If you are still seeing this message please open an issue. logger=ngalert input="foo=" origin=config err="end of input: expected label value" suggestion="foo=\"\""
WARN [02-02|13:18:27] Matchers input has disagreement          logger=ngalert input="baz=\"\\xf0\\x9f\\x99\\x82\"" origin=config
```

Unlike notification policies, silences do not cause warnings, as the Silence API uses matchers as JSON, meaning no parsing is required:

Screenshot 2024-02-02 at 1 24 40 PM
Screenshot 2024-02-02 at 1 25 13 PM

**Comment by grobinson-grafana** [comment]
I just had to rebase `main` due to a conflict in `go.mod`.

**Review by yuri-tceretian — APPROVED** [review]
LGTM.

File: pkg/services/ngalert/metrics/alertmanager.go [hunk 1]
```diff
@@ -18,7 +18,7 @@ func NewAlertmanagerMetrics(r prometheus.Registerer) *Alertmanager {
 	other := prometheus.WrapRegistererWithPrefix(fmt.Sprintf("%s_%s_", Namespace, Subsystem), r)
 	return &Alertmanager{
 		Registerer:                r,
-		Alerts:                    metrics.NewAlerts("grafana", other),
+		Alerts:                    metrics.NewAlerts(other),
```
[the remark this answers is not in the document]
Comment (grobinson-grafana) [reply pkg/services/ngalert/metrics/alertmanager.go:21]: I was removed [here](https://github.com/prometheus/alertmanager/commit/b4f7027908402551a2a84dae2f7889bf0ffa8219) when APIv1 was deprecated.

**Comment by grobinson-grafana** [comment]
This pull request updates our fork of Alertmanager to commit 65bdab0, which is based on commit 5658f8c in Prometheus Alertmanager.

It applies the changes from grafana/alerting#155 which removes the overrides for validation of alerts, labels and silences that we had put in place to allow alerts and silences to work for non-Prometheus datasources. However, as this is now supported in Alertmanager with the UTF-8 work, we can use the new upstream functions and remove these overrides.

The compat package is a package in Alertmanager that takes care of backwards compatibility when parsing matchers, validating alerts, labels and silences. It has three modes: classic mode, UTF-8 strict mode, fallback mode. These modes are controlled via compat.InitFromFlags. Grafana initializes the compat package without any feature flags, which is the equivalent of fallback mode. Classic and UTF-8 strict mode are used in Mimir.

While Grafana Managed Alerts have no need for fallback mode, Grafana can still be used as an interface to manage the configurations of Mimir Alertmanagers and view configurations of Prometheus Alertmanager, and those installations might not have migrated or being running on older versions. Such installations behave as if in classic mode, and Grafana must be able to parse their configurations to interact with them for some period of time. As such, Grafana uses fallback mode until we are ready to drop support for outdated installations of Mimir and the Prometheus Alertmanager.

**Pull request merged**